File: deep_learn_opt/Adam.py
from typing import Callable, Dict, List
import numpy as np
import pandas as pd

# Import my functionalities:
import sys

sys.path.append("../useful_python_scripts/")
from useful_plots import plot_model_comparison
from derivative_funcitons import gradient_of_fun_in_point


def get_data() -> List[List]:
    # Read the CSV file
    df = pd.read_csv("data.csv")
    # Extract each column as a vector
    input_signal = df["input"].tolist()
    output_signal = df["output"].tolist()
    # Return the data
    del df
    return [input_signal, output_signal]


def evaluate_output_using_model(
    system_model: List[float], input_signal: List[float], initial_condition: float
) -> List[float]:
    # get the order
    order: int = int(len(system_model) / 2)
    system_model = np.array(system_model)
    # First "order" values assume steady state with initial condition
    extended_output: List[float] = order * [initial_condition]
    extended_input: List[float] = (order * [input_signal[0]]) + input_signal
    for n in np.arange(order, len(extended_input)):
        # Create the window that affects the output at n
        interest_window: List[float] = []
        for i in np.arange(order):
            interest_window.append(-extended_output[n - i - 1])
        for i in np.arange(order):
            interest_window.append(extended_input[n - i - 1])

        # The n-th value of the output is
        extended_output.append(np.dot(system_model, interest_window))

    return extended_output[order:]


def calculate_cuadratic_mean_error(
    output_signal_model: List[float], real_output: List[float]
) -> float:
    cuadratic_mean_error: float = 0.0
    N: int = len(real_output)
    # Acumulate the Cuadratic mean error
    for n in np.arange(N):
        cuadratic_mean_error = cuadratic_mean_error + (
            ((real_output[n] - output_signal_model[n]) ** 2) / N
        )
    # Return the cuadratic mean error
    return cuadratic_mean_error


def loss_function(model: List[float]) -> float:
    # Evaluate output with model
    global x_data, y_data
    output_signal_model = evaluate_output_using_model(model, x_data, y_data[0])
    # Return the cuadratic mean error
    return calculate_cuadratic_mean_error(output_signal_model, y_data)


def Adam_method(
    epsilon: float,
    initial_theta: List[float],
    obj_loss_function: Callable[[List[float]], float],
    max_iter: int = 100,
) -> Dict:
    # For learning
    step_size = 0.001       # step size (called "E" in book)
    p1 = 0.9                # Exponential decay rate p1
    p2 = 0.999              # Exponential decay rate p2
    sigma = 1e-8            # Small constant used to stabilize division by small numbers  

    # Initializations for algorithm
    theta = np.array(initial_theta)     # initial theta 
    s = np.zeros(len(theta))            # initial s
    r = np.zeros(len(theta))            # initial r
    t: int = 0                          # time step

    # Numerical Initializations
    grad_magnitude: float = 1000 * epsilon      # just to start the loop
    iterations = 0 
    best_theta = np.copy(theta)         # to save the best theta
    mini_batch_size: int = 200
    num_batches: int = int(len(input_signal) / mini_batch_size)  
    global x_data, y_data
    
    # Start loop
    while (grad_magnitude > epsilon) and (iterations < max_iter):
        iterations = iterations + 1
        for k in np.arange(num_batches):
            # 1. get minibatch
            x_data = input_signal[k * mini_batch_size : (k + 1) * mini_batch_size]
            y_data = output_signal[k * mini_batch_size : (k + 1) * mini_batch_size] 
            # 2. approximate gradient of loss_function at theta_weird
            grad = gradient_of_fun_in_point(obj_loss_function, theta, epsilon/10) 
            grad_magnitude = np.linalg.norm(grad)
            # 3. update time
            t = t + 1
            # 4. Update biased ﬁrst moment estimate 
            s = p1*s + (1-p1)*grad
            # 5. Update biased second moment estimate
            r = p2*r + ( (1 - p2)*(grad * grad) )
            # 6. Correct bias in ﬁrst moment
            s_hat = s / (1 - (p1**t))
            # 7. Correct bias in second moment
            r_hat = r / (1 - (p2**t))
            # 8. Compute update: 
            delta_theta = - step_size * ( s_hat / (np.sqrt(r_hat) + sigma) )
            # 9. Apply update
            theta = theta + delta_theta 
            # 10. get the best theta calculated
            x_data = input_signal.copy()
            y_data = output_signal.copy() 
            if obj_loss_function(theta) < obj_loss_function(best_theta):
                best_theta = np.copy(theta) 
            # 11. update for new iteration
            iterations = iterations + 1
            if iterations > max_iter:
                iterations = iterations - 1
                break 

    return {"value": list(best_theta), "iterations": iterations}


def main() -> None:
    # Define initial model
    # initial_model = [-1.50679337, 0.57994944, -0.04300483, 0.11451339]
    # initial_model = [-1.506, 0.579, -0.043, 0.114]
    initial_model = [-1.4, 0.6, -0.1, 0.2]
    # initial_model = [0.1, 0.1, -0.1, -0.1]
    print(f"initial model: {initial_model}")
    # Call function to run Stochastic gradient descent (SGD)
    solution: Dict = Adam_method(
        0.001, initial_model, loss_function, 2000
    )
    sol = solution["value"]
    print(f"Solution achieved: {sol}") 
    print(f"function at that value: {loss_function(sol)}. Itereations required: {solution["iterations"]}")

    output_signal_model = evaluate_output_using_model(
        sol, input_signal, output_signal[0]
    )

    # plot the model to compare
    plot_model_comparison(input_signal, output_signal, output_signal_model)


if __name__ == "__main__":
    # Get data to compare
    input_signal, output_signal = get_data()
    # for minibatches
    x_data = input_signal
    y_data = output_signal
    # Run the main function
    main()

File: miniproyect/matlab/adam/adam.m
function solution = adam(epsilon, initial_theta, obj_loss_function, max_iter)

    if nargin < 4
        max_iter = 150;
    end

    % For learning
    step_size = 0.05;       % step size (called "E" in book)
    p1 = 0.9;                % Exponential decay rate p1
    p2 = 0.999;              % Exponential decay rate p2
    sigma = 1e-8;            % Small constant used to stabilize division by small numbers 

    % Initializations for algorithm
    theta = initial_theta;               % initial theta 
    s = zeros(size(theta));            % initial s
    r = zeros(size(theta));            % initial r
    t = 0;                               % time step

    % Numerical Initializations
    grad_magnitude = 1000 * epsilon;      % just to start the loop
    iterations = 0; 
    best_theta = theta;                   % to save the best theta 
    best_obj_loss = obj_loss_function(best_theta);
    
    trayectory = cell(max_iter,1);
    loss_func_trayectory = zeros(size(trayectory));
    % Start loop
    while (grad_magnitude > epsilon) && (iterations < max_iter)
        iterations = iterations + 1; 
        % 1. get minibatch 
        % 2. approximate gradient of loss_function at theta_weird
        fprintf("Iter %d: Calculating gradient",iterations)
        grad = gradient_of_fun_in_point(obj_loss_function, theta, epsilon/10)';
        grad_magnitude = norm(grad);
        % 3. update time
        t = t + 1;
        % 4. Update biased ﬁrst moment estimate 
        s = p1*s + (1-p1)*grad;
        % 5. Update biased second moment estimate
        r = p2*r + ( (1 - p2)*(grad .* grad) );
        % 6. Correct bias in ﬁrst moment
        s_hat = s / (1 - (p1^t));
        % 7. Correct bias in second moment
        r_hat = r / (1 - (p2^t));
        % 8. Compute update: 
        delta_theta = - step_size * ( s_hat ./ (sqrt(r_hat) + sigma) )
        % 9. Apply update
        theta = theta + delta_theta 
        % 10. get the best theta calculated 
        new_obj_loss = obj_loss_function(theta)
        if new_obj_loss < best_obj_loss
            best_theta = theta;  
            best_obj_loss = new_obj_loss;
        end
        % 11. update for new iteration
        trayectory{iterations} = theta;
        loss_func_trayectory(iterations) = new_obj_loss;
        if iterations > max_iter
            iterations = iterations - 1;
            break  
        end  
    end 
    if iterations < max_iter
        converged = true;
    else
        converged = false;
    end
    solution = struct("iterations", iterations, "value", best_theta, "loss_func_value", best_obj_loss,"trayectory", {trayectory}, "converged", converged);
end
 

function gradient = gradient_of_fun_in_point(fun, point, epsilon)
    % Approximate the value of the gradient of a function at a given point using central difference method.
    
    if nargin < 3
        epsilon = 1e-6;
    end
    
    % Define the dimensions
    num_dim = length(point);
    
    % Define a unitary vector in each dimension
    directions = eye(num_dim);
    
    % Calculate the gradient by differentiating in each direction
    gradient = arrayfun(@(i) derivate_fun_in_point_and_dir(fun, point, directions(:, i), epsilon), 1:num_dim);
    % for i = 1:num_dim
    %     gradient(i) = derivate_fun_in_point_and_dir(fun, point, directions(:, i), epsilon);
    % end
end

function d = derivate_fun_in_point_and_dir(fun, point, direction, epsilon)
    % Approximate the value of the derivative of a function at a given point in a given direction using central difference method.
    
    if nargin < 4
        epsilon = 1e-6;
    end
    
    % Check the direction has unitary magnitude
    if norm(direction) ~= 1.0
        direction = (1 / norm(direction)) * direction;
    end
    
    point_left = point - (epsilon * direction);
    point_right = point + (epsilon * direction);
    
    d = (fun(point_right) - fun(point_left)) / (2 * epsilon);
end 